sympy/core/exprtools.py

```diff
@@ -1414,156 +1290,549 @@ def factor_nc(expr):
+                self.rep[degree][term] = new_monom
+        else:
+            self.rep[degree][term] = monom
+
+    def subtract(self, monom):
+        """Subtract a monomial from this polynomial, updating in-place"""
+        return self.add(monom * -1)
+
+
+    def _factor_homogeneous(self, ldegree):
+        """Factor the maximum-degree terms into two polynomials
+        of degree ldegree and self.degree - ldegree, if possible.
+
+        Return (None, None) if no such factorization exists.
+
+        Implementation based on Algorithm 1 from https://arxiv.org/abs/1002.3180
+        """
+        rdegree = self.degree - ldegree
+        terms = list(self.monomials(self.degree))
+        Ghat, Hhat = terms[0].split(ldegree)
+        G = _NCPoly(None, self.generators)
+        H = _NCPoly(None, self.generators)
+        G.add(Ghat)
+        H.add(Hhat)
+        for monomial in terms[1:]:
+            R = monomial.ldivide(Ghat)
+            if R is None:
+                L = monomial.rdivide(Hhat)
+                if L is not None:
+                    G.add(L)
+            else:
+                H.add(R)
+        if len(G) * len(H) != len(terms):
+            return None, None
+        GH = G * H
+        if GH.as_expr() == Add(*[term.as_expr() for term in terms]):
+            return G, H
+        return None, None
+
+
+    def _factor_once(self, ldegree):
+        """Factor into pairs of polynomials of degree ldegree and self.degree - ldegree, if possible.
+        Coefficients will not be factored or otherwise simplified.
+
+        Return a generator that yields tuples of polynomials.
+
+        Implementation based on Algorithm 2 from https://arxiv.org/abs/1002.3180
+        """
+        from sympy.solvers.polysys import solve_poly_system
+        if self.degree == 0:
+            return
+        if len(self) < 2:
+            return
+
+        if ldegree == 0 or ldegree >= self.degree:
+            return
+            
+        G, H = self._factor_homogeneous(ldegree)
+        if G is None:
+            return
+
+        n = self.degree
+        h = ldegree
+        k = n - ldegree
+        # Future optimization: select the monomials with the least overlap potential
+        Ghat = next(G.monomials())
+        Hhat = next(H.monomials())
+        
+        dummies = []
+        Fhat = self.copy()
+
+        for j in range(1, n + 1):
+            for i in range(1, j):
+                Fhat -= G.copy(h - i) * H.copy(k - j + i)
+
+            if j <= min(h, k):
+                Ghat_L, Ghat_R = Ghat.split(h - j)
+                Hhat_L, Hhat_R = Hhat.split(j)
+                if Ghat_R.term == Hhat_L.term:
+                    # Lines 7-8 of the published algorithm suggest that the term should be
+                    # Ghat_L.term * Hhat_R.term. This doesn't make sense, however - the term is
+                    # supposed to have degree n - j, not n - 2j.
+                    term = _NCMonomial._mul_terms(Ghat.term, Hhat_R.term)
```

Comment:
What about the symmetric counterpart Ghat_L and Hhat? Do they always to Ghat and Hhat_R?

Reply:
Okay, it looks like they are same because it is GHat_L * Ghat_R * Hhat_L *Hhat_R with duplicate terms in the middle cancelled